src/cdomains/intDomain.ml

```diff
@@ -1826,9 +1901,483 @@ module Enums : S with type int_t = BigInt.t = struct
+  let (=:) = Ints_t.equal
+  let (<:) x y = Ints_t.compare x y < 0
+  let (>:) x y = Ints_t.compare x y > 0
+  let (<=:) x y = Ints_t.compare x y <= 0
+  let (>=:) x y = Ints_t.compare x y >= 0
+  (* a divides b *)
+  let ( |: ) a b = (b %: a) =: Ints_t.zero
+
+  let to_yojson t = failwith "to yojson unimplemented"
+
+  let rec gcd x y = (*TODO Double Check*)
+    if y =: Ints_t.zero then x else gcd y (x %: y)
+
+  let abs x = let r = Ints_t.neg x
+    in if x <: r then r else x
+
+  let normalize x =
+    match x with
+    | None -> None
+    | Some (c, m) -> if m =: Ints_t.zero then Some (c, m) else Some (c %: (abs m), (abs m))
+
+  let min_int ik = Ints_t.of_bigint @@ fst @@ Size.range_big_int ik
+
+  let max_int ik = Ints_t.of_bigint @@ snd @@ Size.range_big_int ik
+  let top () = Some (Ints_t.zero, Ints_t.one)
+  let top_of ik = Some (Ints_t.zero, Ints_t.one)
+  let bot () = None
+  let bot_of ik = bot () (* TODO: improve  *)
+
+  let is_top x = x = top ()
+
+  let show = function ik -> match ik with
+    | None -> "∅"
+    | Some (c, m) when (c, m) = (Ints_t.zero, Ints_t.zero) -> Ints_t.to_string c
+    | Some (c, m) ->
+      let a = if c =: Ints_t.zero then "" else Ints_t.to_string c in
+      let b = if m =: Ints_t.zero then "" else if m = Ints_t.one then "ℤ" else Ints_t.to_string m^"ℤ" in
+      let c = if a = "" || b = "" then "" else "+" in
+      a^c^b
+
+  let equal a b = match (normalize a), (normalize b) with
+    | None, None -> true
+    | Some (a, b), Some (c, d) -> a =: c && b =: d
+    | _, _ -> false
+
+  include Std (struct type nonrec t = t let name = name let top_of = top_of let bot_of = bot_of let show = show let equal = equal end)
+
+  let equal_to i = function
+    | None -> failwith "unsupported: equal_to with bottom"
+    | Some (a, b) when b =: Ints_t.zero -> if a = i then `Eq else `Neq
+    | Some (a, b) -> if i %: b = a then `Top else `Neq
+
+  let set_overflow_flag ik =
+    if Cil.isSigned ik && !GU.in_verifying_stage then
+      Goblintutil.did_overflow := true
+
+  let norm ik = function None -> None | Some (c,m) ->
+    if c <: min_int ik   || c >: max_int ik || m <: min_int ik || m >: max_int ik then (set_overflow_flag ik; top_of ik)
+    else Some (c,m)
+
+  let leq (x:t) (y:t) =
+    match x, y with
+    | None, _ -> true
+    | Some _, None -> false
+    | Some (c1,m1), Some (c2,m2) when m2 =: Ints_t.zero && m1 =: Ints_t.zero -> c1 =: c2
+    | Some (c1,m1), Some (c2,m2) when m2 =: Ints_t.zero -> c1 =: c2 && m1 =: Ints_t.zero
+    | Some (c1,m1), Some (c2,m2) -> m2 |: (gcd (c1 -: c2) m2)
+
+  let leq x y =
+    let res = leq x y in
+    if M.tracing then M.trace "congruence" "leq %a %a -> %a \n" pretty x pretty y pretty (Some(Ints_t.of_int (Bool.to_int res), Ints_t.zero)) ;
+    res
+
+  let join ik (x:t) y =
+    match x, y with
+    | None, z | z, None -> z
+    | Some (c1,m1), Some (c2,m2) ->
+      let m3 = gcd m1 (gcd m2 (c1 -: c2)) in
+      norm ik @@ normalize (Some (c1, m3))
```

Comment:
what is the difference between `norm` and `normalize`. They should be renamed to more obvious names